### ISSUE TITLE
Segmentation fault on SELECT after vaccuming database

### ISSUE THREAD

**Issue by kolbasky** [timescale/timescaledb#3163]
We have a zabbix installation, that uses PostgreSQL as a backend database.
Decided to install timescaledb extension.
We have
PostgreSQL 12.6 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44), 64-bit
timescaledb-2-loader-postgresql-12.x86_64    2.2.0-0.el7                    @timescale_timescaledb
timescaledb-2-postgresql-12.x86_64           2.2.0-0.el7                    @timescale_timescaledb
timescaledb-tools.x86_64                     0.11.0-0.el7                   @timescale_timescaledb

we have 3 PG nodes in patroni cluster

So we installed extension, following timescaledb manual.
Ran timescaledb_tune, applying all the settings. Except for wal_min_size.
We stopped zabbix and killed all sessions from it.
Then executed database.sql from zabbix which is:
SELECT create_hypertable('history', 'clock', chunk_time_interval => 86400, migrate_data => true);
SELECT create_hypertable('history_uint', 'clock', chunk_time_interval => 86400, migrate_data => true);
SELECT create_hypertable('history_log', 'clock', chunk_time_interval => 86400, migrate_data => true);
SELECT create_hypertable('history_text', 'clock', chunk_time_interval => 86400, migrate_data => true);
SELECT create_hypertable('history_str', 'clock', chunk_time_interval => 86400, migrate_data => true);
SELECT create_hypertable('trends', 'clock', chunk_time_interval => 2592000, migrate_data => true);
SELECT create_hypertable('trends_uint', 'clock', chunk_time_interval => 2592000, migrate_data => true);
UPDATE config SET db_extension='timescaledb',hk_history_global=1,hk_trends_global=1;
UPDATE config SET compression_status=1,compress_older='7d';

It took a while, but it ran successfully.
After it database works fine for a while.
Then we run vacuumdb -z -F -j 4 zabbix

after a while the server crashes with following errors:
2021-04-26 20:27:50.308 MSK [228295]: [u= h= db= app=] LOG:  automatic analyze of table "zabbix.public.application_discovery" system usage: CPU: user: 0.01 s, system: 0.00 s, elapsed: 0.03 s
2021-04-26 20:27:51.030 MSK [116564]: [u= h= db= app=] LOG:  server process (PID 213417) was terminated by signal 11: Segmentation fault
2021-04-26 20:27:51.030 MSK [116564]: [u= h= db= app=] DETAIL:  Failed process was running: select distinct eventid from event_suppress where eventid=41700297
2021-04-26 20:27:51.030 MSK [116564]: [u= h= db= app=] LOG:  terminating any other active server processes

after it server restarts and crashes with the same error on the same select statement. it is like a boot loop, except that server starts up and crashes right after it.

We restored from backup, ran vacuumdb without any errors, then tried again.
But after installing extension and running database.sql we still face the same issue.

The interesting part is that event_suppress table is not a hypertable, but the error starts happening only after extension installation.

**Comment by kolbasky** [comment]
even more. i tried restoring from backup and just installing extension without running database.sql and running vacuumdb.
we get the same error.

**Comment by mkindahl** [comment]
@kolbasky Thanks for reporting the issue. You don't happen to have a backtrace available? It would simplify finding the problem, or see if we have already fixed it.

**Comment by kolbasky** [comment]
Could you help us with getting backtrace? 
I tried using gdb, but it complaints about missing debug symbols. gdb provides a command to install them, but after installation of debug symbols it still complains about them. Plus, i don't get it - i have to specify pid number to gdb, but i don't know which pid will cause the error.

Other manulas on the internet state that i have to compile postgres with debug symbols, which is 'alittlebita' overkill.

**Comment by mkindahl** [comment]
@kolbasky Normally, you need to have debug symbols, yes, and compiling the code is one way to get it there. Note that there can be debug symbols for both PostgreSQL and for TimescaleDB. It is sufficient if I can see the part of the stack involving TimescaleDB, even if it complains about PostgreSQL symbols. Is that the situation you're in?

**Comment by svenklemm** [comment]
@kolbasky do you still get the segfault after `set timescaledb.enable_skipscan to false;`

**Comment by kolbasky** [comment]
> @kolbasky do you still get the segfault after `set timescaledb.enable_skipscan to false;`

@svenklemm sorry for late reply - i had to arrange a test environment, to let our production system work.

Yes, it completely fixed the issue. And the problem turned out to be not about vacuum, but about distinct selects.
Could you, please, explain what are the consequences of disabling this option? Is it safe to use in production?

**Comment by mkindahl** [comment]
@kolbasky With 2.2 we have introduced an optimization called skip scan that generates special plans for `SELECT DISTICT`. This option disabled that plan rewrite. If you want to compare the two plans, you can use an `EXPLAIN` before the query and see that if skip scan is enabled, there is a custom scan node present called `SkipScan`.

We have fixed a number of issues in Skip Scan in the upcoming 2.2.1 release. Is it possible for you to add the result of `EXPLAIN VERBOSE` on your query so that we can see if it is resolve by any of those fixes?

**Comment by victor-sudakov** [comment]
I seem to be having the same problem on PostgreSQL 12.6-1.pgdg100+1, timescaledb-2-postgresql-12 2.2.0~debian10. The following query

`select distinct graphid from graphs_items where graphid=8842 and not itemid in (66228, 66239,66250,66261,66272,66283,66294);
`

segfaults the PostgreSQL process with 100% reproducibility. Setting `timescaledb.enable_skipscan=off ` fixes the crash.

**Comment by mkindahl** [comment]
@victor-sudakov Thank you for the plan after disabling skip scan. Do you have the explain for before as well?

**Comment by victor-sudakov** [comment]
> @victor-sudakov Thank you for the plan after disabling skip scan. Do you have the explain for before as well?

Sorry, cannot possibly produce it. The EXPLAIN for before had crashed the PostgreSQL process exactly the way the real query crashed it.

**Comment by kolbasky** [comment]
the same for me.

**Comment by mkindahl** [comment]
@victor-sudakov @kolbasky That is strange, this means that there is a bug in the generation of plans, not in the execution. (Did you accidentally include `ANALYZE`? That would run the query and could trigger a crash during execution.) Do either of you have a table definition with all indexes on that table (in particular, the index `graph_items_2`)?

**Comment by kolbasky** [comment]
@mkindahl  i have run it without ANALYZE.
Here are the definitions of two tables. Since they are different in my and victor's situations:

**zabbix=# \d+ event_suppress**
```
                                   Table "public.event_suppress"
      Column      |  Type   | Collation | Nullable | Default | Storage | Stats target | Description
------------------+---------+-----------+----------+---------+---------+--------------+-------------
 event_suppressid | bigint  |           | not null |         | plain   |              |
 eventid          | bigint  |           | not null |         | plain   |              |
 maintenanceid    | bigint  |           |          |         | plain   |              |
 suppress_until   | integer |           | not null | 0       | plain   |              |
Indexes:
    "event_suppress_pkey" PRIMARY KEY, btree (event_suppressid)
    "event_suppress_1" UNIQUE, btree (eventid, maintenanceid)
    "event_suppress_2" btree (suppress_until)
    "event_suppress_3" btree (maintenanceid)
Foreign-key constraints:
    "c_event_suppress_1" FOREIGN KEY (eventid) REFERENCES events(eventid) ON DELETE CASCADE
    "c_event_suppress_2" FOREIGN KEY (maintenanceid) REFERENCES maintenances(maintenanceid) ON DELETE CASCADE
Access method: heap
```

**zabbix=# \d+ graphs_items**
```
                                                  Table "public.graphs_items"
  Column   |         Type         | Collation | Nullable |           Default           | Storage  | Stats target | Description
-----------+----------------------+-----------+----------+-----------------------------+----------+--------------+-------------
 gitemid   | bigint               |           | not null |                             | plain    |              |
 graphid   | bigint               |           | not null |                             | plain    |              |
 itemid    | bigint               |           | not null |                             | plain    |              |
 drawtype  | integer              |           | not null | 0                           | plain    |              |
 sortorder | integer              |           | not null | 0                           | plain    |              |
 color     | character varying(6) |           | not null | '009600'::character varying | extended |              |
 yaxisside | integer              |           | not null | 0                           | plain    |              |
 calc_fnc  | integer              |           | not null | 2                           | plain    |              |
 type      | integer              |           | not null | 0                           | plain    |              |
Indexes:
    "graphs_items_pkey" PRIMARY KEY, btree (gitemid)
    "graphs_items_1" btree (itemid)
    "graphs_items_2" btree (graphid)
Foreign-key constraints:
    "c_graphs_items_1" FOREIGN KEY (graphid) REFERENCES graphs(graphid) ON DELETE CASCADE
    "c_graphs_items_2" FOREIGN KEY (itemid) REFERENCES items(itemid) ON DELETE CASCADE
Access method: heap
```

**Comment by victor-sudakov** [comment]
> Did you accidentally include `ANALYZE`? 

I'm sure I did not.

>  Do either of you have a table definition with all indexes on that table (in particular, the index `graph_items_2`)?
```

> \d+ graphs_items
                                                  Table "public.graphs_items"
  Column   |         Type         | Collation | Nullable |           Default           | Storage  | Stats target | Description 
-----------+----------------------+-----------+----------+-----------------------------+----------+--------------+-------------
 gitemid   | bigint               |           | not null |                             | plain    |              | 
 graphid   | bigint               |           | not null |                             | plain    |              | 
 itemid    | bigint               |           | not null |                             | plain    |              | 
 drawtype  | integer              |           | not null | 0                           | plain    |              | 
 sortorder | integer              |           | not null | 0                           | plain    |              | 
 color     | character varying(6) |           | not null | '009600'::character varying | extended |              | 
 yaxisside | integer              |           | not null | 0                           | plain    |              | 
 calc_fnc  | integer              |           | not null | 2                           | plain    |              | 
 type      | integer              |           | not null | 0                           | plain    |              | 
Indexes:
    "graphs_items_pkey" PRIMARY KEY, btree (gitemid)
    "graphs_items_1" btree (itemid)
    "graphs_items_2" btree (graphid)
Foreign-key constraints:
    "c_graphs_items_1" FOREIGN KEY (graphid) REFERENCES graphs(graphid) ON DELETE CASCADE
    "c_graphs_items_2" FOREIGN KEY (itemid) REFERENCES items(itemid) ON DELETE CASCADE
Access method: heap

\d+ graphs_items_2 
                 Index "public.graphs_items_2"
 Column  |  Type  | Key? | Definition | Storage | Stats target 
---------+--------+------+------------+---------+--------------
 graphid | bigint | yes  | graphid    | plain   | 
btree, for table "public.graphs_items"


```

**Comment by svenklemm** [comment]
@kolbasky This should be fixed in 2.2.1. Feel free to reopen if you can reproduce the issue with 2.2.1.

Duplicate of #3146

**Comment by victor-sudakov** [comment]
After upgrading to 2.2.1, I reset timescaledb.enable_skipscan to on, ran the query, no crash anymore.
